build/patches/experimental-user-scripts-support.patch

```diff
@@ -0,0 +1,10299 @@
++import android.net.Uri;
++import androidx.annotation.Nullable;
++
++import org.chromium.base.annotations.CalledByNative;
++import org.chromium.base.annotations.JNINamespace;
++import org.chromium.base.annotations.NativeMethods;
++import org.chromium.base.Log;
++import org.chromium.ui.base.WindowAndroid;
++
++import org.chromium.components.user_scripts.ScriptListPreference;
++import org.chromium.components.user_scripts.IUserScriptsUtils;
++
++@JNINamespace("user_scripts")
++public class UserScriptsBridge {
++    static WeakReference<ScriptListPreference> observer;
++
++    private static IUserScriptsUtils utilInstance;
++
++    public static void registerUtils(IUserScriptsUtils instance) {
++        utilInstance = instance;
++    }
++
++    public static IUserScriptsUtils getUtils() {
++        return utilInstance;
++    }
++
++    public static boolean isFeatureEnabled() {
++        return UserScriptsBridgeJni.get().isFeatureEnabled();
++    }
++
++    public static boolean isUserEnabled() {
++        return UserScriptsBridgeJni.get().isUserEnabled();
++    }
++
++    public static void setUserEnabled(boolean enabled) {
++        UserScriptsBridgeJni.get().setUserEnabled(enabled);
++    }
++
++    public static void RemoveScript(String key) {
++        UserScriptsBridgeJni.get().removeScript(key);
++    }
++
++    public static void SetScriptEnabled(String key,
++                                        boolean enabled) {
++        UserScriptsBridgeJni.get().setScriptEnabled(key, enabled);
++    }
++
++    public static void Reload() {
++        UserScriptsBridgeJni.get().reload();
++    }
++
++    public static void SelectAndAddScriptFromFile(WindowAndroid window) {
++        Context context = window.getContext().get();
++
++        Intent fileSelector = new Intent(Intent.ACTION_OPEN_DOCUMENT);
++        fileSelector.addCategory(Intent.CATEGORY_OPENABLE);
++        fileSelector.setType("*/*");
++        fileSelector.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
++
++        window.showIntent(fileSelector,
++            new WindowAndroid.IntentCallback() {
++                @Override
++                public void onIntentCompleted(WindowAndroid window, int resultCode, Intent data) {
++                    if (data == null) return;
++                    Uri filePath = data.getData();
++                    UserScriptsBridgeJni.get().tryToInstall(filePath.toString());
++                }
++            },
++            null);
++    }
++
++    public static void TryToInstall(String ScriptFullPath) {
++        UserScriptsBridgeJni.get().tryToInstall(ScriptFullPath);
++    }
++
++    public static List<ScriptInfo> getUserScriptItems() {
++        List<ScriptInfo> list = new ArrayList<>();
++        try {
++            String json = UserScriptsBridgeJni.get().getScriptsInfo();
++
++            JSONObject jsonObject = new JSONObject(json);
++
++            JSONArray scripts = jsonObject.names();
++            if (scripts != null) {
```

Comment:
When can this be null? Maybe we should log in that case to help the user troubleshoot?

Reply:
I don't remember, I think when the list is empty, I check

Reply:
added log in https://github.com/bromite/bromite/pull/857/commits/8c164138391082350f20493a0ca5c4926887372c